enhancements/machine-config/onprem-namespaces.md

```diff
@@ -0,0 +1,134 @@
+---
+title: namespaces-for-on-prem-networking-services
+authors:
+  - "@cybertron"
+reviewers:
+  - "@smarterclayton"
+  - "@celebdor"
+  - "@mandre"
+  - "@rgolangh"
+  - "@patrickdillon"
+  - "@kikisdeliveryservice"
+approvers:
+  - Someone from MCO
+  - Someone from architecture
+creation-date: 2020-07-24
+last-updated: 2020-07-27
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Namespaces for On-Prem Networking Services
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+When support for on-premise deployments on baremetal, OpenStack, OVirt, and
+VSphere was added, some additional networking services were needed to provide
+functionality that would come from cloud services on other platforms. These
+services were added in namespaces called openshift-kni-infra,
+openshift-openstack-infra, etc. There are two problems with this: 1) The
+namespaces are created in all deployments, including on platforms other
+than the one that uses the namespace and 2) The openshift-x-infra name scheme
+does not match the existing pattern for OpenShift namespaces.
+
+## Motivation
+
+The existence of namespaces not related to a given deployment is confusing to
+users. Additionally, the inconsistent naming of the namespaces is a further
+possible point of confusion.
+
+### Goals
+
+- Only create namespaces for networking services on deployments that need them.
+- Use names that are consistent with the rest of OpenShift.
+
+### Non-Goals
+
+## Proposal
+
+Instead of creating the namespaces at install time as is currently done, add
+them to the list of
+[resources to be synced](https://github.com/openshift/machine-config-operator/blob/master/pkg/operator/sync.go)
+in the machine-config-operator. This will allow us to add logic that considers
+the platform before creating them.
+
```

Comment:
So static pods are used to provide these services right? and these static pods require the namespace to exist before they are run right? What about service accounts and rbac i'm hoping the same thing right?

If we are depending on machine-config-operator sync.go that means it is going to create and install the namespaces when it starts running, but to run the operator you need the control-plane nodes. So i think we need to document how this will be handled during bootstrapping too as that is important part of the flow.

Reply:
Also some of these static pods also run on bootstrap host and the requirement of namespaces, rbac etc would be true, how do we plan to handle that.

Reply:
I'm not sure they do require the namespace before they run. I don't entirely understand how, but according to the commit message on https://github.com/openshift/machine-config-operator/commit/b3ab7d8cfdbdd717bfefee62a56c6a1d0b03d657 we were previously running these pods without the namespace. Apparently the only issue was accounting for the resource usage on the nodes.

I also don't believe we have any service accounts or rbac for these. Any interaction with the api is done via the host's kubeconfig.